### PR TITLE
SSR: Add support for media, sizes and heights attribute

### PR DESCRIPTION
**Pull request by sebastianbenz**
This will make it possible to remove the boilerplate even if the `media`, `sizes` or `heights` attributes are used. These attributes will be transformed into CSS media queries instead.

### CODE REVIEW

**Comment by sebastianbenz** [comment]
@westonruter @schlessera PTAL - I haven't yet implemented the CSS size check.

File: packages/optimizer/lib/transformers/ApplyCommonAttributes.js [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+const parseSizes = require('../parseSizes');
+const {appendChild, createElement, insertText, hasAttribute} = require('../NodeUtils');
+const ID_PREFIX = 'i-amp-';
```
Comment (westonruter) [inline packages/optimizer/lib/transformers/ApplyCommonAttributes.js:20]: The prefix `i-amphtml-` is disallowed, but `i-amp-` is OK. It seems somewhat strange why `i-amp-` wasn't chosen rather for the internal prefix since it has fewer bytes.

File: packages/optimizer/lib/transformers/ApplyCommonAttributes.js [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+      // The user agent will pick a width from the sizes attribute, using the first item with a <media-condition> (the part in parentheses) that evaluates to true.
+      // This means, we have to reverse the order the media queries in CSS to emulate this behavior (the last definition has precedence).
+      heights: heights.values.reverse(),
+    });
+  }
+
+  toString() {
+    let result = '';
+    for (const {heights, defaultHeight, id} of this.heights) {
+      const selector = `#${id}:first-child`;
```
Comment (westonruter) [inline packages/optimizer/lib/transformers/ApplyCommonAttributes.js:159]: Shouldn't this be:

```suggestion
      const selector = `#${id}>:first-child`;
```

See https://github.com/ampproject/amp-wp/pull/4482#discussion_r421605106

File: packages/optimizer/lib/transformers/ApplyCommonAttributes.js [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+      return;
+    }
+    if (!customStyles) {
+      customStyles = createElement('style', {
+        'amp-custom': '',
+      });
+      appendChild(head, customStyles);
+      insertText(customStyles, '');
+    }
+    customStyles.children[0].data += styles;
```
Comment (westonruter) [inline packages/optimizer/lib/transformers/ApplyCommonAttributes.js:230]: As noted in the description, we'll need to make sure this doesn't go above 75000 bytes.

File: packages/optimizer/lib/transformers/ApplyCommonAttributes.js [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+   * Returns an existing id or generates a new one.
+   *
+   * @param {Node} node
+   */
+  getOrCreateId(node) {
+    if (hasAttribute(node, 'id')) {
+      return node.attribs.id;
+    }
+    node.attribs = node.attribs || [];
+    node.attribs.id = ID_PREFIX + this.counter;
```
Comment (westonruter) [inline packages/optimizer/lib/transformers/ApplyCommonAttributes.js:248]: Just in case…

```suggestion
    node.attribs.id = ID_PREFIX + this.counter.toString();
```

File: packages/optimizer/spec/transformers/valid/ServerSideRendering/converts_heights_attribute_to_css/input.html [hunk 5]
```diff
@@ -2,32 +2,27 @@
   <head>
     <script async src=https://cdn.ampproject.org/v0.js></script>
     <script async custom-element=amp-img src=https://cdn.ampproject.org/v0/amp-img-0.1.js></script>
+    <noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
     <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;
 -moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes-amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes-amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes-amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes-amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style>
-    <noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
```
Comment (schlessera) [inline packages/optimizer/spec/transformers/valid/ServerSideRendering/converts_heights_attribute_to_css/input.html:7]: Why the change in order here?

File: packages/optimizer/spec/transformers/valid/ServerSideRendering/converts_media_attribute_to_css/expected_output.html [hunk 6]
```diff
@@ -0,0 +1,28 @@
+<html ⚡ i-amphtml-layout i-amphtml-no-boilerplate>
+<head><style amp-runtime></style>
+  <script async src="https://cdn.ampproject.org/v0.js"></script>
+  <script async custom-element="amp-img" src="https://cdn.ampproject.org/v0/amp-img-0.1.js"></script>
+
+
```
Comment (schlessera) [inline packages/optimizer/spec/transformers/valid/ServerSideRendering/converts_media_attribute_to_css/expected_output.html:6]: Removal of the boilerplate shouldn't leave empty lines behind.

File: packages/optimizer/lib/parseSizes.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+let Size;
+
+/**
+ * Parses the text representation of sizes into Sizes object.
+ * See https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#Attributes.
+ * See https://html.spec.whatwg.org/multipage/embedded-content.html#attr-img-sizes
+ * See https://html.spec.whatwg.org/multipage/images.html#parsing-a-sizes-attribute
+ *
+ * @param {String} string
+ * @returns {!Sizes}
```
Comment (schlessera) [inline packages/optimizer/lib/parseSizes.js:43]: Why the "non-nullable" modifier `!` here specifically? Wouldn't that be the default without the modifier anyways? It seems like that distinction isn't use anywhere else.

File: packages/optimizer/lib/transformers/ApplyCommonAttributes.js [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+ * Transforms a media attribute into CSS by negating the existing media query.
+ */
+class MediaTransformer {
+  constructor() {
+    this.media = new Map();
+  }
+
+  transform(node, id) {
+    // normalize whitespace
+    let mediaString = node.attribs.media.replace(/\s+/g, ' ');
```
[the remark this answers is not in the document]
Comment (sebastianbenz) [reply packages/optimizer/lib/transformers/ApplyCommonAttributes.js:32]: parseSizes is unrelated to the media attribute or am I missing something?

File: packages/optimizer/lib/parseSizes.js (hunk 7 above)
Comment (sebastianbenz) [reply packages/optimizer/lib/parseSizes.js:43]: It's a closure compiler flag used by the AMP runtime. Removed it as we don't need it.

File: packages/optimizer/lib/transformers/ApplyCommonAttributes.js [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+      }
+    }
+  }
+
+  /**
+   * Returns an existing id or generates a new one.
+   *
+   * @param {Node} node
+   */
+  getOrCreateId(node) {
```
[the remark this answers is not in the document]
Comment (sebastianbenz) [reply packages/optimizer/lib/transformers/ApplyCommonAttributes.js:243]: True. This is not straightforward to implement in our case unfortunately.

File: packages/optimizer/lib/transformers/ApplyCommonAttributes.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (schlessera) [reply packages/optimizer/lib/transformers/ApplyCommonAttributes.js:32]: Oh, nvm, I was thinking of the media conditions that are part of the `sizes` attribute.

File: packages/optimizer/spec/transformers/valid/ServerSideRendering/converts_media_attribute_to_css/expected_output.html (hunk 6 above)
Comment (sebastianbenz) [reply packages/optimizer/spec/transformers/valid/ServerSideRendering/converts_media_attribute_to_css/expected_output.html:6]: that's going to be removed by a different transformer (not used in the test)

File: packages/optimizer/spec/transformers/valid/ServerSideRendering/converts_heights_attribute_to_css/input.html (hunk 5 above)
Comment (sebastianbenz) [reply packages/optimizer/spec/transformers/valid/ServerSideRendering/converts_heights_attribute_to_css/input.html:7]: reverted

File: packages/optimizer/lib/transformers/ApplyCommonAttributes.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (sebastianbenz) [reply packages/optimizer/lib/transformers/ApplyCommonAttributes.js:243]: I'll ignore this case for now. Re-running the transformer would simply re-use the existing id. Pre-existing Ids should not be avoided by our prefix.

**Comment by sebastianbenz** [comment]
I've addressed the comments. In regards to checking for the CSS limit. I'm currently lending towards not including this check:

* CSS optimization would need to be run twice (before, to be able to get the correct byte count and after, to optimize the generated CSS rules).
* I'd prefer to fail on not being able to remove the boilerplate, the perf implications of not being able to remove the boilerplate because of this are a lot harder to detect otherwise.

**Comment by sebastianbenz** [comment]
@westonruter @schlessera PTAL

**Comment by westonruter** [comment]
>  In regards to checking for the CSS limit. I'm currently lending towards not including this check

Doesn't this run the risk of the Optimizer generating invalid AMP from valid AMP input? If the CSS is close to the limit and then optimization causes it to go over, that seems not ideal?

File: packages/optimizer/spec/transformers/valid/ServerSideRendering/converts_media_attribute_to_css/expected_output.html (hunk 6 above)
Comment (sebastianbenz) [reply packages/optimizer/spec/transformers/valid/ServerSideRendering/converts_media_attribute_to_css/expected_output.html:6]: This will be fixed by a different transformer

File: packages/optimizer/lib/transformers/ApplyCommonAttributes.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (sebastianbenz) [reply packages/optimizer/lib/transformers/ApplyCommonAttributes.js:264]: Done

**Comment by sebastianbenz** [comment]
Merging as there's no further feedback that needs to be addressed.

**Comment by schlessera** [comment]
@sebastianbenz There seems to be an issue with the CSS in here that was raised twice but always resolved without a comment, so I assume that GitHub resolved it because of an unrelated change: https://github.com/ampproject/amp-toolbox/pull/763#discussion_r425816316

Or am I missing something maybe?

**Comment by sebastianbenz** [comment]
Hah! Not sure how this got lost. Fixed in #811

**Comment by sebastianbenz** [comment]
Thanks for picking up on this!
